core/domain/stats_jobs_one_off.py

```diff
@@ -28,14 +33,224 @@
+                       'version': item.exp_version,
+                       'state_name': item.state_name,
+                       'is_feedback_useful': item.is_feedback_useful
+                   })
+        elif isinstance(item, stats_models.StateHitEventLogEntryModel):
+            yield (item.exploration_id,
+                   {
+                       'event_type': feconf.EVENT_TYPE_STATE_HIT,
+                       'version': item.exploration_version,
+                       'state_name': item.state_name,
+                       'session_id': item.session_id
+                   })
+        elif isinstance(item, stats_models.SolutionHitEventLogEntryModel):
+            yield (item.exp_id,
+                   {
+                       'event_type': feconf.EVENT_TYPE_SOLUTION_HIT,
+                       'version': item.exp_version,
+                       'state_name': item.state_name,
+                       'session_id': item.session_id
+                   })
+        elif isinstance(
+                item, stats_models.ExplorationActualStartEventLogEntryModel):
+            yield (item.exp_id,
+                   {
+                       'event_type': feconf.EVENT_TYPE_ACTUAL_START_EXPLORATION,
+                       'version': item.exp_version,
+                       'state_name': item.state_name
+                   })
+        elif isinstance(
+                item, stats_models.CompleteExplorationEventLogEntryModel):
+            yield (item.exploration_id,
+                   {
+                       'event_type': feconf.EVENT_TYPE_COMPLETE_EXPLORATION,
+                       'version': item.exploration_version,
+                       'state_name': item.state_name
+                   })
+
+    @staticmethod
+    def reduce(exp_id, values):
+        values = map(ast.literal_eval, values)
+        values = sorted(values, key=lambda x: x['version'])
+
+        # Find the latest version number
+        exploration = exp_services.get_exploration_by_id(exp_id)
+        latest_exp_version = exploration.version
+        versions = range(1, latest_exp_version + 1)
+
+        # Get a copy of the corrupted statistics models to copy uncorrupted
+        # v1 fields
+        old_stats = stats_services.get_multiple_exploration_stats_by_version(
+            exp_id, versions)
+        # Get list of snapshot models for each version of the exploration
+        snapshots_by_version = (
+            exp_models.ExplorationModel.get_snapshots_metadata(
+                exp_id, versions))
+
+        exp_stats_dicts = []
+        event_idx = 0
+        event = values[event_idx]
+        for version in versions:
+            datastore_stats_for_version = old_stats[version-1]
+            if version == 1:
+                # Reset the possibly corrupted stats
+                datastore_stats_for_version.num_starts_v2 = 0
+                datastore_stats_for_version.num_completions_v2 = 0
+                datastore_stats_for_version.num_starts_v2 = 0
```

Comment:
This copies line 136, should it be actual_starts instead?

Reply:
Yes I think it should be. Done.